### PR TITLE
Adding support and test unit for "EXTENDED_DRX_PARAMETERS" NAS PDU IE

### PR DESCRIPTION
**Pull request by alksara**
## Summary

This PR resolves attach reject - cause 99 (NAS PDU Information Element non-existent or not implemented) issue here (https://github.com/magma/magma/issues/8430). It adds support for Extended_DRX_Paramaters NAS PDU IE sent by some terminals(e.g. Alcatel Linkkey LTE cat 4 usb dongle).

Changes made:
1- Added decoding support for optional IE ("EXTENDED_DRX_PARAMETERS") in AttachRequest.c 
2- Declared "EXTENDED_DRX_PARAMETERS" in AttachRequest.h
3- Added test unit for testing decoding of "EXTENDED_DRX_PARAMETERS" in test_mme_app_emm_decode.cpp

## Test Plan

1. Tested the changes against test units in "test_mme_app_emm_decode.cpp". 
Proc:
->ran "make test_oai" under "~/magma/lte/gateway" on vagrant dev. VM--> all tests passed

```
vagrant@magma-dev-focal:~/magma/lte/gateway$ make test_oai
OAI_FLAGS -DS6A_OVER_GRPC=True -DEMBEDDED_SGW=True -DENABLE_OPENFLOW=True -DSPGW_ENABLE_SESSIOND_AND_MOBILITYD=True
mkdir -p   /home/vagrant/build/c/core
cd   /home/vagrant/build/c/core &&  cmake   /home/vagrant/magma/lte/gateway/c/core -DCMAKE_BUILD_TYPE=Debug   -DS6A_OVER_GRPC=True -DEMBEDDED_SGW=True -DENABLE_OPENFLOW=True -DSPGW_ENABLE_SESSIOND_AND_MOBILITYD=True -DBUILD_TESTS=1 -GNinja
-- Configuring done
-- Generating done
-- Build files have been written to: /home/vagrant/build/c/core
# Turn on generation of compile_commands.json (See GH Issue #5488)
cd   /home/vagrant/build/c/core && awk '{if (/^CMAKE_EXPORT_COMPILE_COMMANDS/) gsub(/OFF/, "ON"); print}' CMakeCache.txt > tmp && mv tmp CMakeCache.txt
ninja -C   /home/vagrant/build/c/core
ninja: Entering directory `/home/vagrant/build/c/core'
[0/1] Re-running CMake...
-- Configuring done
-- Generating done
-- Build files have been written to: /home/vagrant/build/c/core
[1/7] Performing build step for 'MagmaCommon'
ninja: no work to do.
[4/7] Performing configure step for 'MagmaCore'
Build type is RelWithDebInfo
Building with sentry!
Proto_srcs are /home/vagrant/build/c/core/oai/common/lte/protos/oai/common_types.pb.cc;/home/vagrant/build/c/core/oai/common/orc8r/protos/common.pb.cc;/home/vagrant/build/c/core/oai/common/orc8r/protos/redis.pb.cc;/home/vagrant/build/c/core/oai/common/lte/protos/mconfig/mconfigs.pb.cc
Proto_srcs are /home/vagrant/build/c/core/oai/lib/sgs_client/feg/protos/csfb.pb.cc;/home/vagrant/build/c/core/oai/lib/sgs_client/feg/protos/csfb.grpc.pb.cc;/home/vagrant/build/c/core/oai/lib/sgs_client/lte/protos/mconfig/mconfigs.pb.cc
Proto_srcs are /home/vagrant/build/c/core/oai/lib/sgs_client/feg/protos/csfb.pb.cc;/home/vagrant/build/c/core/oai/lib/sgs_client/feg/protos/csfb.grpc.pb.cc;/home/vagrant/build/c/core/oai/lib/sgs_client/lte/protos/mconfig/mconfigs.pb.cc;/home/vagrant/build/c/core/oai/lib/sgs_client/orc8r/protos/common.pb.cc
Proto_srcs are /home/vagrant/build/c/core/oai/lib/sms_orc8r_client/lte/protos/sms_orc8r.pb.cc;/home/vagrant/build/c/core/oai/lib/sms_orc8r_client/lte/protos/sms_orc8r.grpc.pb.cc;/home/vagrant/build/c/core/oai/lib/sms_orc8r_client/lte/protos/mconfig/mconfigs.pb.cc
Proto_srcs are /home/vagrant/build/c/core/oai/tasks/nas/lte/protos/oai/nas_state.pb.cc;/home/vagrant/build/c/core/oai/tasks/nas/lte/protos/oai/spgw_state.pb.cc;/home/vagrant/build/c/core/oai/tasks/nas/lte/protos/oai/common_types.pb.cc;/home/vagrant/build/c/core/oai/tasks/nas/lte/protos/oai/std_3gpp_types.pb.cc
s8 Proto_srcs are /home/vagrant/build/c/core/oai/tasks/sgw_s8/lte/protos/oai/spgw_state.pb.cc;/home/vagrant/build/c/core/oai/tasks/sgw_s8/lte/protos/oai/sgw_state.pb.cc;/home/vagrant/build/c/core/oai/tasks/sgw_s8/lte/protos/oai/common_types.pb.cc;/home/vagrant/build/c/core/oai/tasks/sgw_s8/lte/protos/oai/std_3gpp_types.pb.cc
s8 Proto_includes are /home/vagrant/build/c/core/oai/tasks/sgw_s8/lte/protos/oai/spgw_state.pb.h;/home/vagrant/build/c/core/oai/tasks/sgw_s8/lte/protos/oai/sgw_state.pb.h;/home/vagrant/build/c/core/oai/tasks/sgw_s8/lte/protos/oai/common_types.pb.h;/home/vagrant/build/c/core/oai/tasks/sgw_s8/lte/protos/oai/std_3gpp_types.pb.h
-- Configuring done
-- Generating done
-- Build files have been written to: /home/vagrant/build/c/core/oai
[5/7] Performing build step for 'MagmaCore'
ninja: no work to do.
[7/7] Completed 'MagmaCore'
cd  /home/vagrant/build/c/core/oai && ctest --output-on-failure
Test project /home/vagrant/build/c/core/oai
    Start 1: test_mme_app_ue_context
1/8 Test #1: test_mme_app_ue_context ..........   Passed    0.03 sec
    Start 2: test_mme_app_emm_decode
2/8 Test #2: test_mme_app_emm_decode ..........   Passed    0.02 sec
    Start 3: test_openflow_controller
3/8 Test #3: test_openflow_controller .........   Passed    0.02 sec
    Start 4: test_imsi_encoder
4/8 Test #4: test_imsi_encoder ................   Passed    0.01 sec
    Start 5: test_gtp_app
5/8 Test #5: test_gtp_app .....................   Passed    0.02 sec
    Start 6: test_spgw_state_converter
6/8 Test #6: test_spgw_state_converter ........   Passed    0.01 sec
    Start 7: test_itti
7/8 Test #7: test_itti ........................   Passed    4.22 sec
    Start 8: test_pipelined_client
8/8 Test #8: test_pipelined_client ............   Passed    0.01 sec

100% tests passed, 0 tests failed out of 8

Total Test time (real) =   4.35 sec
```

2. Tested the changes on deployment AGW connected to Mavenir AP and Alcatel Linkkey LTE cat 4 usb dongle. 
Proc:
-->Built MME binary on Vagrant dev. VM by running "make BUILD_TYPE=RelWithDebInfo build_oai" under "~/magma/lte/gateway" , transferred binary to Deployment VM, backed up mme binary, stopped magma services (sudo service magma@* stop),  replaced mme binary with the new one, restarted  magma services (sudo service magma@magmad restart) --> all services came up, UE successfully got attached and connected. Issue seen here (https://github.com/magma/magma/issues/8430) resolved. 

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: lte/gateway/c/core/oai/tasks/nas/emm/msg/AttachRequest.c [hunk 1]
```diff
@@ -338,6 +338,15 @@ int decode_attach_request(
         decoded += 1;  // Device Properties is 1 byte
         break;
 
+      case ATTACH_REQUEST_EXTENDED_DRX_PARAMETERS_IEI:
+        // Skip this IE.
+        OAILOG_INFO(
+            LOG_NAS_EMM,
+            "EMM-MSG - Extended DRX Parameters IE in Attach Request is not "
+            "supported. Skipping this IE.");
+        decoded += 3;  // Extended DRX Parameters is 3 bytes
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/tasks/nas/emm/msg/AttachRequest.c:347]: This IE is a type 4 information element. You need to implement a `decode_extended_drx_parameter_ie` function that checks the existence of IEI (1 octet), then read the length (1 octet) and then skip the remaining value part based on its length read in the previous octet.

File: lte/gateway/c/core/oai/test/mme_app_task/test_mme_app_emm_decode.cpp [hunk 2]
```diff
@@ -67,6 +67,28 @@ TEST_F(EMMDecodeTest, TestDecodeAttachRequest2) {
   bdestroy_wrapper(&attach_request.esmmessagecontainer);
 }
 
+TEST_F(EMMDecodeTest, TestDecodeAttachRequest3) {
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/test/mme_app_task/test_mme_app_emm_decode.cpp:70]: Make the test name more informative e.g., `TestDecodeAttachRequestWithExtDrx`.

Comment (ulaskozat) [reply lte/gateway/c/core/oai/test/mme_app_task/test_mme_app_emm_decode.cpp:70]: Add multiple test cases to this test with different lengths for extended drx IE.

File: lte/gateway/c/core/oai/tasks/nas/emm/msg/AttachRequest.c (hunk 1 above)
Comment (ulaskozat) [reply lte/gateway/c/core/oai/tasks/nas/emm/msg/AttachRequest.c:347]: you can check `decode_voice_domain_preference_and_ue_usage_setting` to see a sample.

**Review by ulaskozat — CHANGES_REQUESTED** [review]
- The IE that is decoded is of type 4  (TLV or LV). You need to write the decoder and encode functions accordingly. 
- I also suggest that you actually decode and store the values of Paging Time Window and eDRX values as uint8_t in a new `e_drx_parameter_t` field in EMM context (check out `drx_parameter_t`).
- Make the test name more informing and cover test cases with different length values for this IE.

**Comment by ulaskozat** [comment]
@alksara Also you should first raise the PR against the master branch and then we backport merged PR to v1.6. Otherwise it will create unexpected cases for next version does not support the feature while the older version does support it.

**Comment by ulaskozat** [comment]
@alksara do you plan to update your PR?

**Comment by alksara** [comment]
@ulaskozat yes, was on vacation last week, and got busy with other stuff this week. will work on it.

**Comment by ekowtaylor** [comment]
What are the plans to move this forward?

**Comment by ekowtaylor** [comment]
@alksara any interest in rebasing and resubmitting for review?

**Pull request closed without merging**